### ISSUE TITLE
Question: Should analyzers for VB.NET be written in VB.NET?

### ISSUE THREAD

**Issue by zspitz** [dotnet/vblang#360]
### Pro - C#:

1. Wider pool of potential maintainers (but see point 3)
2. Pattern matching (as of this writing)
3. Extending an existing analyzer (but see point 4)

### Pro VB.NET

3. VB.NET programmers are more familiar with the language, and are thus better suited to write an analyzer.
4. Analyzers don't really lend themselves to cross-language extension -- the `SyntaxKind` enums and the syntax node classes are different between languages.

The specific analyzer that I want to write for VB.NET already exists for C# [here](https://github.com/Microsoft/RoslynClrHeapAllocationAnalyzer), with a [request for VB.NET support](https://github.com/Microsoft/RoslynClrHeapAllocationAnalyzer/issues/62).

ping @AnthonyDGreen @bandleader

**Comment by zspitz** [comment]
@reduckted It's interesting to note that even in the UseNameofInPlaceOfString analyzer, the only language-specific parts are to figure out if `NameOf` exists in the current version. The actual logic for the analyzer and the code fix is shared across languages..

**Comment by zspitz** [comment]
@reduckted From [.NET Compiler Platform - Language-Agnostic Code Generation with Roslyn](https://msdn.microsoft.com/en-us/magazine/mt707527.aspx):

> You can use the Roslyn APIs and the `SyntaxGenerator` class whenever you need to perform rich analysis over source code, but this approach is also very useful with analyzers and code refactorings. In fact, analyzers, code fixes, and refactorings have the `DiagnosticAnalyzer`, `ExportCodeFixProvider`, and `ExportCodeRefactoringProvider` attributes, respectively, each accepting the primary and secondary supported languages. By using `SyntaxGenerator` instead of `SyntaxFactory`, you can target both C# and Visual Basic simultaneously.

Is this what you meant?

(NB. I've read it a few times by now, but I'm still not sure how to apply it.)

**Comment by paul1956** [comment]
I have written several refactoring and would love to contribute but have no idea how to even start (where are they added , who approved which ones people want, are they part of Roslyn or somewhere else).

The one I think is most useful adds or removes type information from Declares, based on Option settings. They are all written in VB though I do depend on one C# module to deal with Hash that I just cloned out of Roslyn. I have been able to process all of Roslyn if you don't mind waiting a few hours I have not updated the recently so I don't know if performance is improved with newer Roslyn libraries.
```VB
Dim X = 1
' Becomes
Dim X as Integer = 1
' Becomes
Dim X = 1
```

**Comment by zspitz** [comment]
@paul1956 The directions [here](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix) seem straightforward. The 'Analyzer with code fix" template produces a .vsix, which can then be published to the Marketplace as described [here](https://docs.microsoft.com/en-us/visualstudio/extensibility/walkthrough-publishing-a-visual-studio-extension?view=vs-2017#package-your-extension). Once you upload to the Marketplace, your analyzer will be available to anyone to install into VS.

Alternatively, since the template creates a .NET Standard DLL, you can upload it as a Nuget package, as described [here](https://docs.microsoft.com/en-us/nuget/quickstart/create-and-publish-a-package-using-visual-studio). Then, your projects can add it as a NuGet reference.

(NB. I haven't actually tried either of these; I'm still trying to wrap my head around writing the analyzer code.)

**Comment by paul1956** [comment]
@zspitz So these are not open source on GitHub? I have written both the VSIX kind and NuGet package you get both from the template, don't know how the NuGet one helps anyone. Right now my VSIX has 20+ VB analyzers/Codefixes all written in VB but some are only interesting to me or are available from others like code-cracker, would I just package the one that does the adding and removing Types from DIm statements and Function/Sub declarations?

**Comment by zspitz** [comment]
@paul1956 

> these are not open source on GitHub

I'm not sure what you mean. You don't have to publish your analyzer to GitHub or any other open-source repository. If it's only for your personal use, then you may not need to publish it at all; just add a reference to the built DLL.

But I think other people could use the analyzer you're describing, so you might want to consider packaging it separately and publishing it.

> don't know how the NuGet one helps anyone

If I'm not mistaken, once it's in a NuGet package, you can either add it to a [local feed](https://docs.microsoft.com/en-us/nuget/hosting-packages/local-feeds), or [upload to nuget.org](https://docs.microsoft.com/en-us/nuget/quickstart/create-and-publish-a-package-using-visual-studio). Then, it can be found by using the NuGet package management in VS (in the case of a local feed, you have to [add the local feed](https://docs.microsoft.com/en-us/nuget/tools/package-manager-ui#package-sources)).

The analyzer will then be available to any project that references it, but only to that project.

If you want it available to every project, then it needs to be packaged as a VSIX.

**Comment by jrmoreno1** [comment]
I am attempting to write an Analyzer/CodeFixProvider in C# for VB (pattern matching is awesome, although I think #172 would have worked just as well if I was writing vb), and while I can run and debug the analyzer, I am unable to get the CodeFixProvider to be executed when debugging (haven’t tried it in a non-debug session yet).

When trying to add the vb nuget packages for code analysis it looked like it was restricting the choices to c# packages somehow, had to do things manually once before I could get the vb dll’s for the analyzer.

**Comment by jrmoreno1** [comment]
My guidance from CyrusNajmabadi on the linked question was to use VB for a vb Analyzer.

**Comment by CyrusNajmabadi** [comment]
Yes. In general roslyn writes vb analyzers/fixers in VB. Though if we write something for both c# and vb then we'll usually extract all the shared logic into code in a common lib.
